snapcraft/extractors/appstream.py

```diff
@@ -53,13 +55,29 @@ def extract(path: str) -> ExtractedMetadata:
         # --elopio -20180109
         icon = node.text.strip()
 
-    desktop_file_ids = []
+    desktop_file_paths = []
     nodes = tree.findall('launchable')
     for node in nodes:
         if ('type' in node.attrib and
                 node.attrib['type'] == 'desktop-id'):
-            desktop_file_ids.append(node.text.strip())
+            desktop_file_id = node.text.strip()
+            desktop_file_path = _desktop_file_id_to_path(desktop_file_id)
+            if desktop_file_path:
+                desktop_file_paths.append(desktop_file_path)
 
     return ExtractedMetadata(
         summary=summary, description=description, icon=icon,
-        desktop_file_ids=desktop_file_ids)
+        desktop_file_paths=desktop_file_paths)
+
+
+def _desktop_file_id_to_path(desktop_file_id: str) -> str:
+    # For details about desktop file ids and their corresponding paths, see
+    # https://standards.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html#desktop-file-id  # noqa
+    for xdg_data_dir in ('usr/local/share', 'usr/share'):
```

Comment:
Should this also take `share`? Snaps built from source may not use "usr" and that's where they end up by default when using eg. the CMake plugin.

Reply:
I would say no. If the desktop file id doesn't follow the xdg rules, it is wrong. If we go further and add our own heuristics to find the files, it would be confusing.
Another option that we discuss was to try to find the files in all the directories. I would prefer that over adding more paths not in the standard.

Reply:
Was there a reason not to go through all directories? Like .desktop files not being used for launchers? My main concern is that the layout in the snap doesn't necessarily look the same as in a Debian package and it may not be obvious to the user of Snapcraft where files will be found.

Reply:
You can always specify the path on the snapcraft.yaml, using the `desktop` attribute. My thought here was to first do just the common and standard cases, not more. Then we can add complexity, doing a walk through all the files in pull, build or prime, if for some reason we require it.